src/Microsoft.Identity.Web/IncrementalConsentAndConditionalAccessHelper.cs

```diff
@@ -11,90 +11,90 @@
+			// When calling an [Authenticate]-decorated controller we expect an authenticated
+			// user and therefore its account should be in the cache. However in the case of an
+			// InMemoryCache, the cache could be empty if the server was restarted. This is why
+			// the null_user exception is thrown.
+			return ex.ErrorCode.ContainsAny(new[] { MsalError.UserNullError, MsalError.InvalidGrantError });
+		}
 
-        /// <summary>
-        /// Build authentication properties needed for incremental consent.
-        /// </summary>
-        /// <param name="scopes">Scopes to request.</param>
-        /// <param name="ex"><see cref="MsalUiRequiredException"/> instance.</param>
-        /// <param name="user">User.</param>
-        /// <param name="userflow">Userflow being invoked for AAD B2C.</param>
-        /// <returns>AuthenticationProperties.</returns>
-        public static AuthenticationProperties BuildAuthenticationProperties(
-            string[]? scopes,
-            MsalUiRequiredException ex,
-            ClaimsPrincipal user,
-            string? userflow = null)
-        {
-            if (ex == null)
-            {
-                throw new ArgumentNullException(nameof(ex));
-            }
+		/// <summary>
+		/// Build authentication properties needed for incremental consent.
+		/// </summary>
+		/// <param name="scopes">Scopes to request.</param>
+		/// <param name="ex"><see cref="MsalUiRequiredException"/> instance.</param>
+		/// <param name="user">User.</param>
+		/// <param name="userflow">Userflow being invoked for AAD B2C.</param>
+		/// <returns>AuthenticationProperties.</returns>
+		public static AuthenticationProperties BuildAuthenticationProperties(
+			string[]? scopes,
+			MsalUiRequiredException ex,
+			ClaimsPrincipal user,
+			string? userflow = null)
+		{
+			if (ex == null)
+			{
+				throw new ArgumentNullException(nameof(ex));
+			}
 
-            scopes ??= new string[0];
-            var properties = new AuthenticationProperties();
+			scopes ??= new string[0];
+			var properties = new AuthenticationProperties();
 
-            // Set the scopes, including the scopes that ADAL.NET / MSAL.NET need for the token cache
-            string[] additionalBuiltInScopes =
-            {
-                 OidcConstants.ScopeOpenId,
-                 OidcConstants.ScopeOfflineAccess,
-                 OidcConstants.ScopeProfile,
-            };
+			// Set the scopes, including the scopes that ADAL.NET / MSAL.NET need for the token cache
+			string[] additionalBuiltInScopes =
+			{
+				 OidcConstants.ScopeOpenId,
+				 OidcConstants.ScopeOfflineAccess,
+				 OidcConstants.ScopeProfile,
+			};
 
-            properties.SetParameter<ICollection<string>>(
-                OpenIdConnectParameterNames.Scope,
-                scopes.Union(additionalBuiltInScopes).ToList());
+			properties.SetParameter<ICollection<string>>(
+				OpenIdConnectParameterNames.Scope,
+				scopes.Union(additionalBuiltInScopes).ToList());
```

Comment:
LINQ is super slow. Like, so slow, I would recommend people almost never use it when there is a non-verbose alternative. In this case, there is a data structure that perfectly does what we want in an optimized way. That data structure is the HashSet<string> which happens to have a UnionWith method and already implements ICollection<string>.